src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

```diff
@@ -103,8 +103,32 @@ public static unsafe void PrepareMethod(RuntimeMethodHandle method, RuntimeTypeH
         [MethodImpl(MethodImplOptions.InternalCall)]
         public static extern void PrepareDelegate(Delegate d);
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int GetHashCode(object? o)
+        {
+            if (o is not null)
+            {
+                ref IntPtr startOfDataRef = ref Unsafe.As<byte, IntPtr>(ref Unsafe.As<RawData>(o).Data);
+                ref IntPtr objectHeaderRef = ref Unsafe.Add(ref startOfDataRef, -2);
```

Comment:
This is a GC hole. This byref will points to previous object. so it won't move together with the object that you are computing the hashcode for.

Reply:
Ouch 😅
We were just about wondering whether the GC could track refs pointing to the object header, I wasn't completely sure but figured it might work given that the data was still part of the same object - guess I know now ahah
I've tried using `fixed` there to fix that but as expected that's pretty slow and loses virtually all performance improvements than the current solution, so not really worth it anymore. Will close the PR for now then.

While on the topic - @SingleAccretion found a `GT_START_NONGC` node in the emitter, and together with @EgorBo we were wondering whether it might make sense and/or be doable at all to introduce a new JIT intrinsic to be able to leverage that? Might make things like this possible without having to pin stuff and lose all performance gains? At the very least it sounds like a good learning opportunity so I thought I'd ask! Thanks! 😄

Reply:
I do not think intrinsics for `GT_START_NONGC` make sense. It is so subtle and hard to get these things right. I have no problems with giving up the bit of performance that we would get.

Reply:
Right, yeah that makes perfect sense and it'd also be extremely niche anyway. Thanks! 🙂